File: src/app/utils/analysisEngine.ts

import { STORY_KEY_POINTS, STORY_TEXT } from "./storyData";
import type { AcousticFeatures } from "./speechPipeline";

// ─── Types ────────────────────────────────────────────────────────────────────

export interface AnalysisResult {
  score: number;
  keyPointsRecalled: number;
  totalKeyPoints: number;
  semanticSimilarity: number;
  coherenceScore: number;
  detailsRecalled: string[];
  detailsMissed: string[];
  riskLevel: "Low" | "Moderate" | "High";
  recommendations: string[];
}

// ─── 1. Levenshtein distance ──────────────────────────────────────────────────

function levenshteinDistance(str1: string, str2: string): number {
  const len1 = str1.length;
  const len2 = str2.length;
  const matrix: number[][] = [];
  for (let i = 0; i <= len1; i++) matrix[i] = [i];
  for (let j = 0; j <= len2; j++) matrix[0][j] = j;
  for (let i = 1; i <= len1; i++) {
    for (let j = 1; j <= len2; j++) {
      const cost = str1[i - 1] === str2[j - 1] ? 0 : 1;
      matrix[i][j] = Math.min(
        matrix[i - 1][j] + 1,
        matrix[i][j - 1] + 1,
        matrix[i - 1][j - 1] + cost
      );
    }
  }
  return matrix[len1][len2];
}

function calculateSimilarity(original: string, recalled: string): number {
  const distance = levenshteinDistance(original.toLowerCase(), recalled.toLowerCase());
  const maxLen = Math.max(original.length, recalled.length);
  return ((maxLen - distance) / maxLen) * 100;
}

// ─── 2. Key-point matching (fuzzy, multi-keyword) ────────────────────────────

const STOP_WORDS = new Set([
  "the", "and", "her", "she", "was", "had", "for", "are", "but", "not",
  "you", "all", "can", "its", "him", "his", "with", "that", "this",
  "from", "they", "have", "been", "were", "said", "each", "will",
]);

function extractKeywords(phrase: string): string[] {
  return phrase
    .toLowerCase()
    .replace(/[()]/g, "")
    .split(/[\s,/]+/)
    .filter((w) => w.length >= 3 && !STOP_WORDS.has(w));
}

function fuzzyContains(haystack: string, needle: string, threshold = 0.72): boolean {
  const needleWords = needle.toLowerCase().split(/\s+/).length;
  const haystackWords = haystack.toLowerCase().split(/\s+/);
  if (haystackWords.length < needleWords) return false;
  for (let i = 0; i <= haystackWords.length - needleWords; i++) {
    const window = haystackWords.slice(i, i + needleWords).join(" ");
    const maxLen = Math.max(window.length, needle.length);
    if (maxLen === 0) continue;
    const dist = levenshteinDistance(window, needle.toLowerCase());
    if ((maxLen - dist) / maxLen >= threshold) return true;
  }
  return false;
}

function checkKeyPoints(recalledText: string): {
  recalled: string[];
  missed: string[];
} {
  const recalled: string[] = [];
  const missed: string[] = [];
  const lower = recalledText.toLowerCase();

  STORY_KEY_POINTS.forEach((point) => {
    const keywords = extractKeywords(point);
    const keywordMatch = keywords.some((kw) => lower.includes(kw));
    const fuzzyMatch = fuzzyContains(lower, point);
    if (keywordMatch || fuzzyMatch) {
      recalled.push(point);
    } else {
      missed.push(point);
    }
  });

  return { recalled, missed };
}

// ─── 3. Coherence scoring ─────────────────────────────────────────────────────

function calculateCoherence(text: string): number {
  if (!text || text.trim().length < 20) return 0;
  const sentences = text.split(/[.!?]+/).filter((s) => s.trim().length > 0);
  if (sentences.length === 0) return 0;
  let score = 0;
  if (text.length > 50)              score += 33;
  if (sentences.length > 2)          score += 33;
  if (text.split(/\s+/).length > 15) score += 34;
  return score;
}

// ─── 4. Rule-based risk classification (fallback when XGBoost unavailable) ────
//
// Thresholds derived from WMS-IV Logical Memory norms for ages 65–89.
// This is used when the XGBoost endpoint returns null (network error,
// model not yet deployed, or environment variables not set).

function classifyRisk(
  immediateScore: number,
  delayedScore: number,
  recallDecay: number
): "Low" | "Moderate" | "High" {
  if (delayedScore < 45 || recallDecay > 25) return "High";
  if (delayedScore >= 70 && recallDecay <= 15) return "Low";
  return "Moderate";
}

// ─── 5. Recommendations ──────────────────────────────────────────────────────

function getRecommendations(risk: "Low" | "Moderate" | "High"): string[] {
  if (risk === "High") {
    return [
      "Consult with a neurologist or geriatric specialist",
      "Consider a comprehensive cognitive assessment (MCI/dementia work-up)",
      "Schedule follow-up testing in 3–6 months",
      "Discuss family history and risk factors with your healthcare provider",
    ];
  }
  if (risk === "Moderate") {
    return [
      "Monitor cognitive function with regular check-ups",
      "Consider lifestyle modifications (diet, exercise, social engagement)",
      "Schedule follow-up screening in 6–12 months",
      "Engage in mentally stimulating activities",
    ];
  }
  return [
    "Continue regular health check-ups",
    "Maintain healthy lifestyle habits",
    "Repeat this screening annually as part of preventive care",
    "Stay mentally and socially active",
  ];
}

// ─── 6. XGBoost prediction ────────────────────────────────────────────────────
//
// Calls the Supabase Edge Function that runs xgboost_model.tsx.
// Returns null (triggering the rule-based fallback above) if:
//   - VITE_SUPABASE_PROJECT_ID or VITE_SUPABASE_ANON_KEY are not set
//   - The endpoint is unreachable
//   - The model has not yet been deployed
//
// Add to your .env file:
//   VITE_SUPABASE_PROJECT_ID=your-project-id
//   VITE_SUPABASE_ANON_KEY=your-anon-key

async function getXGBoostPrediction(
  immediateScore: number,
  delayedScore: number,
  recallDecay: number,
  immediateKP: number,
  delayedKP: number,
  immSimilarity: number,
  delSimilarity: number,
  immCoherence: number,
  delCoherence: number
): Promise<"Low" | "Moderate" | "High" | null> {
  const projectId    = import.meta.env.VITE_SUPABASE_PROJECT_ID as string | undefined;
  const publicAnonKey = import.meta.env.VITE_SUPABASE_ANON_KEY as string | undefined;

  if (!projectId || !publicAnonKey) {
    // Environment variables not configured — skip XGBoost, use rule-based fallback
    return null;
  }

  try {
    const rawAcoustics = sessionStorage.getItem("delayedAcoustics");
    const acoustics: Partial<AcousticFeatures> = rawAcoustics
      ? JSON.parse(rawAcoustics)
      : {};

    const features = {
      pauseCount:         acoustics.pauseCount        ?? null,
      meanPauseDuration:  acoustics.meanPauseDuration ?? null,
      speechRate:         acoustics.speechRate        ?? null,
      lexicalDiversity:   acoustics.lexicalDiversity  ?? null,
      wordsPerMinute:     acoustics.wordsPerMinute    ?? null,
      immediateScore,
      delayedScore,
      recallDecay,
      keyPointsRecalled:  delayedKP,
      semanticSimilarity: delSimilarity,
      coherenceScore:     delCoherence,
    };

    const response = await fetch(
      `https://${projectId}.supabase.co/functions/v1/make-server-57b7b6f3/predict-risk`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${publicAnonKey}`,
        },
        body: JSON.stringify(features),
      }
    );

    if (!response.ok) {
      console.warn("XGBoost endpoint returned", response.status, "— using rule-based fallback");
      return null;
    }

    const data = await response.json();

    if (data.success && data.prediction?.risk_level) {
      // Cache the full prediction for the Results page to display
      sessionStorage.setItem("xgboostPrediction", JSON.stringify(data.prediction));
      return data.prediction.risk_level as "Low" | "Moderate" | "High";
    }

    return null;
  } catch (error) {
    console.warn("XGBoost prediction failed — using rule-based fallback:", error);
    return null;
  }
}

// ─── 7. Main export ───────────────────────────────────────────────────────────

export async function analyzeRecall(
  immediateRecall: string,
  delayedRecall: string
): Promise<{
  immediate: AnalysisResult;
  delayed: AnalysisResult;
  comparison: {
    recallDecay: number;
    concernLevel: "Low" | "Moderate" | "High";
  };
}> {
  // Score immediate recall
  const immediateKP   = checkKeyPoints(immediateRecall);
  const immSimilarity = calculateSimilarity(STORY_TEXT, immediateRecall);
  const immCoherence  = calculateCoherence(immediateRecall);
  const immediateScore =
    (immediateKP.recalled.length / STORY_KEY_POINTS.length) * 40 +
    (immSimilarity / 100) * 30 +
    (immCoherence / 100) * 30;

  // Score delayed recall
  const delayedKP     = checkKeyPoints(delayedRecall);
  const delSimilarity = calculateSimilarity(STORY_TEXT, delayedRecall);
  const delCoherence  = calculateCoherence(delayedRecall);
  const delayedScore =
    (delayedKP.recalled.length / STORY_KEY_POINTS.length) * 40 +
    (delSimilarity / 100) * 30 +
    (delCoherence / 100) * 30;

  const recallDecay = immediateScore - delayedScore;

  // Rule-based classification (always computed as fallback)
  const immediateRisk = classifyRisk(immediateScore, immediateScore, 0);
  const delayedRisk   = classifyRisk(immediateScore, delayedScore, recallDecay);
  let   concernLevel  = classifyRisk(immediateScore, delayedScore, recallDecay);

  // XGBoost prediction — overrides rule-based if the endpoint responds
  const xgboostPrediction = await getXGBoostPrediction(
    immediateScore,
    delayedScore,
    recallDecay,
    immediateKP.recalled.length,
    delayedKP.recalled.length,
    immSimilarity,
    delSimilarity,
    immCoherence,
    delCoherence
  );

  if (xgboostPrediction) {
    concernLevel = xgboostPrediction;
  }

  return {
    immediate: {
      score:              immediateScore,
      keyPointsRecalled:  immediateKP.recalled.length,
      totalKeyPoints:     STORY_KEY_POINTS.length,
      semanticSimilarity: immSimilarity,
      coherenceScore:     immCoherence,
      detailsRecalled:    immediateKP.recalled,
      detailsMissed:      immediateKP.missed,
      riskLevel:          immediateRisk,
      recommendations:    getRecommendations(immediateRisk),
    },
    delayed: {
      score:              delayedScore,
      keyPointsRecalled:  delayedKP.recalled.length,
      totalKeyPoints:     STORY_KEY_POINTS.length,
      semanticSimilarity: delSimilarity,
      coherenceScore:     delCoherence,
      detailsRecalled:    delayedKP.recalled,
      detailsMissed:      delayedKP.missed,
      riskLevel:          delayedRisk,
      recommendations:    getRecommendations(delayedRisk),
    },
    comparison: {
      recallDecay,
      concernLevel,
    },
  };
}

File: utils/analysis_engine.py

from typing import Dict, List, Tuple
from difflib import SequenceMatcher
from utils.story_data import STORY_KEY_POINTS, STORY_TEXT


# ─── Text similarity ──────────────────────────────────────────────────────────

def levenshtein_distance(str1: str, str2: str) -> int:
    len1, len2 = len(str1), len(str2)
    matrix = [[0] * (len2 + 1) for _ in range(len1 + 1)]
    for i in range(len1 + 1):
        matrix[i][0] = i
    for j in range(len2 + 1):
        matrix[0][j] = j
    for i in range(1, len1 + 1):
        for j in range(1, len2 + 1):
            cost = 0 if str1[i - 1] == str2[j - 1] else 1
            matrix[i][j] = min(
                matrix[i - 1][j] + 1,
                matrix[i][j - 1] + 1,
                matrix[i - 1][j - 1] + cost,
            )
    return matrix[len1][len2]


def calculate_similarity(original: str, recalled: str) -> float:
    distance = levenshtein_distance(original.lower(), recalled.lower())
    max_len = max(len(original), len(recalled))
    if max_len == 0:
        return 0.0
    return ((max_len - distance) / max_len) * 100


# ─── Key-point matching (fixed) ───────────────────────────────────────────────

_STOP_WORDS = {
    "the", "and", "her", "she", "was", "had", "for", "are", "but", "not",
    "you", "all", "can", "its", "him", "his", "with", "that", "this",
    "from", "they", "have", "been", "were", "said", "each", "will",
}


def _extract_keywords(phrase: str) -> List[str]:
    """Returns meaningful words from a key-point phrase (length >= 3, no stop words)."""
    tokens = phrase.lower().replace("(", "").replace(")", "").split()
    return [w for w in tokens if len(w) >= 3 and w not in _STOP_WORDS]


def check_key_points(recalled_text: str) -> Tuple[List[str], List[str]]:
    """
    Checks how many story key points appear in the recalled text.

    Uses multi-keyword matching plus a difflib fuzzy fallback so that
    paraphrases (e.g. 'red, yellow, and white' for 'three colors of roses')
    are still credited.
    """
    recalled: List[str] = []
    missed: List[str] = []
    lower_recalled = recalled_text.lower()

    for point in STORY_KEY_POINTS:
        keywords = _extract_keywords(point)
        keyword_match = any(kw in lower_recalled for kw in keywords)

        fuzzy_match = False
        if not keyword_match:
            ratio = SequenceMatcher(None, point.lower(), lower_recalled).ratio()
            fuzzy_match = ratio > 0.40

        if keyword_match or fuzzy_match:
            recalled.append(point)
        else:
            missed.append(point)

    return recalled, missed


# ─── Coherence scoring ────────────────────────────────────────────────────────

def calculate_coherence(text: str) -> float:
    if not text or len(text.strip()) < 20:
        return 0.0

    sentences = [
        s.strip()
        for s in text.replace("!", ".").replace("?", ".").split(".")
        if s.strip()
    ]
    if not sentences:
        return 0.0

    score = 0.0
    if len(text) > 50:
        score += 33.33
    if len(sentences) > 2:
        score += 33.33
    if len(text.split()) > 15:
        score += 33.34

    return score


# ─── Risk classification (rule-based, replaces Decision Tree) ────────────────
#
# Thresholds derived from WMS-IV Logical Memory norms for ages 65–89:
#   Normal delayed recall   : >= 70 % of weighted score
#   Mild impairment (MCI)   : 45–69 %
#   Significant impairment  : < 45 %
#   Normal recall decay     : <= 15 percentage points
#   Concerning decay        : 16–25 pp
#   High decay              : > 25 pp
#
# The XGBoost model (ml_training/train_xgboost.py) will replace this
# function when a validated dataset is available and the model is deployed. 
  
def classify_risk(immediate_score: float, delayed_score: float, recall_decay: float) -> str:
    if delayed_score < 45 or recall_decay > 25:
        return "High"
    if delayed_score >= 70 and recall_decay <= 15:
        return "Low"
    return "Moderate"


# ─── Recommendations ─────────────────────────────────────────────────────────

def get_recommendations(risk: str) -> List[str]:
    if risk == "High":
        return [
            "Consult with a neurologist or geriatric specialist",
            "Consider a comprehensive cognitive assessment (MCI/dementia work-up)",
            "Schedule follow-up testing in 3–6 months",
            "Discuss family history and risk factors with your healthcare provider",
        ]
    if risk == "Moderate":
        return [
            "Monitor cognitive function with regular check-ups",
            "Consider lifestyle modifications (diet, exercise, social engagement)",
            "Schedule follow-up screening in 6–12 months",
            "Engage in mentally stimulating activities",
        ]
    return [
        "Continue regular health check-ups",
        "Maintain healthy lifestyle habits",
        "Repeat this screening annually as part of preventive care",
        "Stay mentally and socially active",
    ]


# ─── Main analysis function ───────────────────────────────────────────────────

def analyze_recall(immediate_recall: str, delayed_recall: str) -> Dict:
    # Immediate recall
    immediate_kp = check_key_points(immediate_recall)
    immediate_similarity = calculate_similarity(STORY_TEXT, immediate_recall)
    immediate_coherence = calculate_coherence(immediate_recall)
    immediate_score = (
        (len(immediate_kp[0]) / len(STORY_KEY_POINTS)) * 40
        + (immediate_similarity / 100) * 30
        + (immediate_coherence / 100) * 30
    )

    # Delayed recall
    delayed_kp = check_key_points(delayed_recall)
    delayed_similarity = calculate_similarity(STORY_TEXT, delayed_recall)
    delayed_coherence = calculate_coherence(delayed_recall)
    delayed_score = (
        (len(delayed_kp[0]) / len(STORY_KEY_POINTS)) * 40
        + (delayed_similarity / 100) * 30
        + (delayed_coherence / 100) * 30
    )

    recall_decay = immediate_score - delayed_score
    risk_level = classify_risk(immediate_score, delayed_score, recall_decay)

    return {
        "immediate": {
            "score": immediate_score,
            "key_points_recalled": len(immediate_kp[0]),
            "total_key_points": len(STORY_KEY_POINTS),
            "semantic_similarity": immediate_similarity,
            "coherence_score": immediate_coherence,
            "details_recalled": immediate_kp[0],
            "details_missed": immediate_kp[1],
            "risk_level": risk_level,
            "recommendations": get_recommendations(risk_level),
        },
        "delayed": {
            "score": delayed_score,
            "key_points_recalled": len(delayed_kp[0]),
            "total_key_points": len(STORY_KEY_POINTS),
            "semantic_similarity": delayed_similarity,
            "coherence_score": delayed_coherence,
            "details_recalled": delayed_kp[0],
            "details_missed": delayed_kp[1],
            "risk_level": risk_level,
            "recommendations": get_recommendations(risk_level),
        },
        "comparison": {
            "recall_decay": recall_decay,
            "concern_level": risk_level,
        },
    }